Firebase/Core/FIRApp.m

```diff
@@ -96,9 +98,9 @@ @implementation FIRApp
 // This is necessary since our custom getter prevents `_options` from being created.
 @synthesize options = _options;
 
-static NSMutableDictionary *sAllApps;
+static GULMutableDictionary *sAllApps;
```

Comment:
Is GULMutableDictionary needed? Would it be sufficient to do the accesses with synchronized(self) like some of the `sAllApps` accesses already are?

Reply:
Just adding `synchronized(self)` would be sufficient. 
My intention here was to fix the current issue and help to prevent it happening in the future. With just `synchronized(self)` in case if new logic is added, a developer would have to remember to add `synchronized(self)`. With `GULMutableDictionary` new logic around `sAllApps` and `sLibraryVersions` will be most likely thread safe without additional efforts.

Please, let me know if I am missing something.

Reply:
It seems like this might be a partial complete fix. I'm a bit worried about destabilizing low-level code that we don't have great tests for. I wonder if today we should do the minimal fix of adding the missing `sychronized(self)`'s and for the next release do the complete fix of everything in this PR **plus** making a new GoogleUtilities subspec **plus** removing the redundant synchronized selfs?

Reply:
Sounds reasonable.
The only thing I am worried about with `sychronized(self)` - I don't know the code well enough to be sure that additional `sychronized(self)` will not introduce a deadlock. Let me take a look at the code one more time.